core/arch/arm/kernel/misc_a32.S

```diff
@@ -21,10 +22,30 @@ END_FUNC __get_core_pos
 /* size_t get_core_pos_mpidr(uint32_t mpidr); */
 FUNC get_core_pos_mpidr , :
 UNWIND(	.fnstart)
-	/* Calculate CorePos = (ClusterId * (cores/cluster)) + CoreId */
-	and	r1, r0, #MPIDR_CPU_MASK
-	and	r0, r0, #MPIDR_CLUSTER_MASK
-	add	r0, r1, r0, LSR #(MPIDR_CLUSTER_SHIFT - CFG_CORE_CLUSTER_SHIFT)
+	mov	r3, r0
+
+	/*
+	 * Shift MPIDR value if it's not already shifted.
+	 * Using logical shift ensures AFF0 to be filled with zeroes.
+	 */
+	tst	r0, #MPIDR_MT_MASK
+	lsleq	r3, r0, #MPIDR_AFFINITY_BITS
```

Comment:
From the ARM ARM I can't tell that shift is needed. As I read it, the MT-bit is only a hint to the scheduler. Please clarify.

Reply:
There's some information about this in the ARM ARM D7.4 section's note (second point).
In practice a non-multi threaded platform's MPIDR values look like `(clusterId <<8) | coreId` and in multi threaded systems they follow the `MT | (clusterId << 16) | (coreId << 8) | threadId` formula. That shift operation is executed if the MT bit was 0 so the following instructions can work on a unified R0 value.

Reply:
Thanks for the pointer. I thought I answered here the other day, but it seems the message got lost somehow.

As I understand it `CFG_THREAD_CORE_SHIFT` must always be correctly defined to match how the CPU is configured so there's no need for semi-dynamic configuration. To keep the code simple and easier to read I would prefer something along:
``` C
#if CFG_THREAD_CORE_SHIFT == 0
	/* Calculate CorePos = (ClusterId * (cores/cluster)) + CoreId */
	and	r1, r0, #MPIDR_AFF0_MASK
	and	r0, r0, #MPIDR_AFF1_MASK
	add	r0, r1, r0, LSR #(MPIDR_AFF1_SHIFT - CFG_CORE_CLUSTER_SHIFT)
#else
	/*
	 * Calculate CorePos =
	 * ((ClusterId * (cores/cluster)) + CoreId) * (threads/core) + ThreadId
	 */
	ubfx	r2, r0, #MPIDR_AFF2_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r1, r0, #MPIDR_AFF1_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r0, r0, #MPIDR_AFF0_SHIFT, #MPIDR_AFFINITY_BITS
	add	r1, r1, r2, LSL #(CFG_CORE_CLUSTER_SHIFT)
	add	r0, r0, r1, LSL #(CFG_THREAD_CORE_SHIFT)
#endif
```
Perhaps and one pair of `ubfx` and `add` instructions can be combined, but it could be worth an extra instruction to keep it more readable.
I think we should use `CFG_CORE_THREAD_SHIFT` instead to harmonize better with all the other `CFG_` variables.
As an extra sanity check we could assert somewhere in `generic_boot.c` that with `CFG_THREAD_CORE_SHIFT` the MT bit isn't set and vice versa.

Reply:
I was a bit misleading in my previous comment because it's a valid setup for single threaded systems to have MPIDR values like this `MT | (clusterId << 16) | (coreId << 8) | 0` (threadId is 0 for all processing elements). So it's a valid case to have the MT bit set beside the CFG_THREAD_CORE_SHIFT is 0. However it's not valid to have the MT bit set, CFG_THREAD_CORE_SHIFT==0 but the AFF0 is **not** 0.
My point is that we need to shift the MPIDR by MPIDR_AFFINITY_BITS if the MT bit is zero to have the unified MPIDR layout before extracting the affinities and use this shifted value in the `#if CFG_THREAD_CORE_SHIFT == 0` block. We could still eliminate two unnecessary instructions. I hope this makes sense.
```c
	tst	r0, #MPIDR_MT_MASK
	lsleq	r3, r0, #MPIDR_AFFINITY_BITS

	/*
	 * At this point the MPIDR layout is always shifted so it looks
	 * as follows AFF2 -> cluster, AFF1 -> core, AFF0 -> thread
	 */
#if CFG_THREAD_CORE_SHIFT == 0
	/* Calculate CorePos = (ClusterId * (cores/cluster)) + CoreId */
	ubfx	r0, r3, #MPIDR_AFF1_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r1, r3, #MPIDR_AFF2_SHIFT, #MPIDR_AFFINITY_BITS
	add	r0, r0, r1, LSL #(CFG_CORE_CLUSTER_SHIFT)
#else
	/*
	 * Calculate CorePos =
	 * ((ClusterId * (cores/cluster)) + CoreId) * (threads/core) + ThreadId
	 */
	ubfx	r0, r3, #MPIDR_AFF0_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r1, r3, #MPIDR_AFF1_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r2, r3, #MPIDR_AFF2_SHIFT, #MPIDR_AFFINITY_BITS
	add	r1, r1, r2, LSL #(CFG_CORE_CLUSTER_SHIFT)
	add	r0, r0, r1, LSL #(CFG_THREAD_CORE_SHIFT)
#endif
```

Reply:
Makes sense, please update the patch. An extra comment explaining what you said above about the MT bit would be nice.

It's likely that some new platforms will copy this function and modify it, so it's important that it's easy to understand how this function works.